pkg/p2p/streamtest/streamtest.go

```diff
@@ -147,12 +155,42 @@ func (s *stream) Write(p []byte) (int, error) {
 }
 
 func (s *stream) Close() error {
+	// lock + closed is used to avoid multiple closing or sending on a closed channel
+	s.mtx.Lock()
+	defer s.mtx.Unlock()
+
+	if s.closed {
+		return nil
+	}
+
 	if err := s.in.Close(); err != nil {
 		return err
 	}
 	if err := s.out.Close(); err != nil {
 		return err
 	}
+
+	select {
+	case s.cin <- struct{}{}:
+		close(s.cin)
+		s.closed = true
+	default:
+	}
```

Comment:
There is no need for select and send on a channel as it is closed right away and only read in FullClose.

The whole block can just be `close(s.cin)`, even without the `closed` variable if `sync.Once` is used.